modules/xmpp/strophe.jingle.js

```diff
@@ -22,6 +22,144 @@ const logger = getLogger(__filename);
 // function call chains.
 /* eslint-disable newline-per-chained-call */
 
+/**
+ * Reads a JSON-encoded message (from a "json-message" element) and extracts source descriptions. Adds the extracted
+ * source descriptions to the given Jingle IQ in the standard Jingle format.
+ *
+ * Encoding sources in this compact JSON format instead of standard Jingle was introduced in order to reduce the
+ * network traffic and load on the XMPP server. The format is described in Jicofo [TODO: insert link].
+ *
+ * @param {*} iq the IQ to which source descriptions will be added.
+ * @param {*} jsonMessageXml The XML node for the "json-message" element.
+ * @returns nothing.
+ */
+function expandSourcesFromJson(iq, jsonMessageXml) {
+
+    let json;
+    try {
+        json = JSON.parse(jsonMessageXml.textContent);
+    } catch (error) {
+        logger.error(`json-message XML contained invalid JSON, ignoring: ${jsonMessageXml.textContent}`);
+        return;
+    }
+
+    if (!json || !json.sources) {
+        // It might be a message of a different type, no need to log.
+        return;
+    }
+
+    // This is where we'll add "source" and "ssrc-group" elements. Create them elements if they don't exist.
+    let videoRtpDescription = getOrCreateRtpDescription(iq, 'video');
+    let audioRtpDescription = getOrCreateRtpDescription(iq, 'audio');
+
+    for (let owner in json.sources) {
+        let ownerSources = json.sources[owner];
+        // The video sources, video ssrc-groups, audio sources and audio ssrc-groups are encoded in that order in the
+        // elements of the array.
+        let videoSources = ownerSources && ownerSources.length > 0 && ownerSources[0];
```

Comment:
What does the json message look like when a given remote endpoint doesn't have video sources but only audio source ? Or when the endpoint doesn't have any video ssrc-groups, like Firefox with simulcast disabled ?